pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go

```diff
@@ -0,0 +1,506 @@
+	} else {
+		glog.V(2).Infof("ensureInternalLoadBalancerDeleted(%v): deleting firewall for healthcheck", loadBalancerName)
+		fwHCName := getFirewallHealthCheckName(loadBalancerName, clusterID, shared)
+		if err = gce.DeleteFirewall(fwHCName); err != nil && !isInUsedByError(err) {
+			return err
+		}
+	}
+
+	// Try deleting instance groups - expect ResourceInuse error if needed by other LBs
+	igName := getInstanceGroupName(clusterID)
+	if err = gce.ensureInstanceGroupsDeleted(igName); err != nil && !isInUsedByError(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (gce *GCECloud) ensureFirewall(fwName, fwDesc string, sourceRanges []string, ports []string, protocol v1.Protocol, nodes []*v1.Node) error {
+	glog.V(2).Infof("ensureFirewall(%v): checking existing firewall", fwName)
+	targetTags, err := gce.GetNodeTags(nodeNames(nodes))
+	if err != nil {
+		return err
+	}
+
+	// Fetch existing firewall
+	existingFirewall, err := gce.GetFirewall(fwName)
+	if err != nil && !isNotFound(err) {
+		return err
+	}
+	// Create expected firewall
+	expectedFirewall := &compute.Firewall{
+		Name:         fwName,
+		Description:  fwDesc,
+		Network:      gce.networkURL,
+		SourceRanges: sourceRanges,
+		TargetTags:   targetTags,
+		Allowed: []*compute.FirewallAllowed{
+			{
+				IPProtocol: strings.ToLower(string(protocol)),
+				Ports:      ports,
+			},
+		},
+	}
+
+	if existingFirewall == nil {
+		glog.V(2).Infof("ensureFirewall(%v): creating firewall", fwName)
+		return gce.CreateFirewall(expectedFirewall)
+	}
+
+	if firewallRuleEqual(expectedFirewall, existingFirewall) {
+		return nil
+	}
+
+	glog.V(2).Infof("ensureFirewall(%v): updating firewall", fwName)
+	return gce.UpdateFirewall(expectedFirewall)
+}
+
+func (gce *GCECloud) ensureFirewalls(loadBalancerName, clusterID string, nm types.NamespacedName, svc *v1.Service, healthCheckPort string, shared bool, nodes []*v1.Node) error {
+	// First firewall is for ingress traffic
+	fwName := loadBalancerName
+	fwDesc := makeFirewallDescription(nm.String(), svc.Spec.LoadBalancerIP)
+	ports, protocol := getPortsAndProtocol(svc.Spec.Ports)
+	sourceRanges, err := v1_service.GetLoadBalancerSourceRanges(svc)
+	if err != nil {
+		return err
+	}
+	err = gce.ensureFirewall(fwName, fwDesc, sourceRanges.StringSlice(), ports, protocol, nodes)
+	if err != nil {
+		return err
+	}
+
+	// Second firewall is for health checking nodes / services
+	fwHCName := getFirewallHealthCheckName(loadBalancerName, clusterID, shared)
+	hcSrcRanges := LoadBalancerSrcRanges()
+	return gce.ensureFirewall(fwHCName, "", hcSrcRanges, []string{healthCheckPort}, v1.ProtocolTCP, nodes)
+}
+
+func (gce *GCECloud) ensureHealthCheck(name, path string, port int32) (*compute.HealthCheck, error) {
+	glog.V(2).Infof("ensureHealthCheck(%v, %v, %v): checking existing health check", name, path, port)
+	wantedHC := newInternalLBHealthCheck(name, path, port)
+
+	hc, err := gce.GetHealthCheck(name)
+	if hc == nil || err != nil && isNotFound(err) {
```

Comment:
This was copied logic from loadbalancer_external.

Separate into two code blocks. One checking for non-NotFound errors; another for handling hc == nil